# Automatically generated by openapi-generator (https://openapi-generator.tech)
# Please update as you see appropriate

context("Test GenomeApi")

api.instance <- GenomeApi$new()

test_that("AssemblyDescriptorsByAccessions", {
  # tests for AssemblyDescriptorsByAccessions
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get genome metadata by accession
  # Get detailed metadata for assembled genomes by accession in a JSON output format.
  # @param array[character]  accessions   
  # @param character  limit  Limit the number of returned results (\"all\", \"none\", otherwise an integer value).  (optional)
  # @param character  filters.refseq.only  If true, only return RefSeq (GCF_) assemblies.  (optional)
  # @param character  returned.content    (optional)
  # @return [V1alpha1AssemblyMetadata]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("AssemblyDescriptorsByTaxon", {
  # tests for AssemblyDescriptorsByTaxon
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get genome metadata by taxonomic identifier
  # Get detailed metadata on all assembled genomes for a specified NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
  # @param character  taxon  NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank 
  # @param character  limit  Limit the number of returned results (\"all\", \"none\", otherwise an integer value).  (optional)
  # @param character  filters.refseq.only  If true, only return RefSeq (GCF_) assemblies.  (optional)
  # @param character  tax.exact.match  If true, only return assemblies with the given NCBI Taxonomy ID, or name. Otherwise, assemblies from taxonomy subtree are included, too. Ignored for assembly_accession request.  (optional)
  # @param character  returned.content    (optional)
  # @return [V1alpha1AssemblyMetadata]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("CheckAssemblyAvailability", {
  # tests for CheckAssemblyAvailability
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Check the validity of genome accessions
  # The &#39;GET&#39; version of check is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param array[character]  accessions  NCBI genome assembly accessions 
  # @param character  include.tsv    (optional)
  # @return [V1alpha1AssemblyDatasetAvailability]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("CheckAssemblyAvailabilityPost", {
  # tests for CheckAssemblyAvailabilityPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Check the validity of many genome accessions in a single request
  # The &#39;GET&#39; version of check is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param V1alpha1AssemblyDatasetRequest  body   
  # @return [V1alpha1AssemblyDatasetAvailability]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadAssemblyPackage", {
  # tests for DownloadAssemblyPackage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a genome dataset by accession
  # Download a genome dataset including fasta sequence, annotation and a detailed data report by accession.
  # @param array[character]  accessions  NCBI genome assembly accessions 
  # @param array[character]  chromosomes  The default setting is all chromosome. Specify individual chromosome by string (1,2,MT or chr1,chr2.chrMT). Unplaced sequences are treated like their own chromosome ('Un'). The filter only applies to fasta sequence.  (optional)
  # @param character  exclude.sequence  Set to true to omit the genomic sequence.  (optional)
  # @param array[character]  include.annotation.type  Select additional types of annotation to include in the data package.  If unset, no annotation is provided.  (optional)
  # @param character  hydrated    (optional)
  # @param character  filename  Output file name.  (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadAssemblyPackagePost", {
  # tests for DownloadAssemblyPackagePost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a genome dataset by post
  # The &#39;GET&#39; version of download is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param V1alpha1AssemblyDatasetRequest  body   
  # @param character  filename  Output file name.  (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDownloadSummary", {
  # tests for GenomeDownloadSummary
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Preview genome dataset download
  # Get a download summary by accession in a JSON output format.
  # @param array[character]  accessions  NCBI genome assembly accessions 
  # @param array[character]  chromosomes  The default setting is all chromosome. Specify individual chromosome by string (1,2,MT or chr1,chr2.chrMT). Unplaced sequences are treated like their own chromosome ('Un'). The filter only applies to fasta sequence.  (optional)
  # @param character  exclude.sequence  Set to true to omit the genomic sequence.  (optional)
  # @param array[character]  include.annotation.type  Select additional types of annotation to include in the data package.  If unset, no annotation is provided.  (optional)
  # @param character  hydrated    (optional)
  # @param character  include.tsv    (optional)
  # @return [V1alpha1DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeDownloadSummaryByPost", {
  # tests for GenomeDownloadSummaryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Preview genome dataset download by POST
  # The &#39;GET&#39; version of download summary is limited by the size of the GET URL (2KB, which works out to about 140 genomic accessions).  The POST operation is provided to allow users to supply a larger number of accessions in a single request.
  # @param V1alpha1AssemblyDatasetRequest  body   
  # @return [V1alpha1DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeTaxNameQuery", {
  # tests for GenomeTaxNameQuery
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a list of taxonomy names and IDs found in the assembly dataset given a partial taxonomic name.
  # This endpoint retrieves a list of taxonomy names and IDs found in the assembly dataset given a partial taxonomic name of any rank.
  # @param character  taxon.query  NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank 
  # @return [V1alpha1SciNameAndIds]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GenomeTaxTree", {
  # tests for GenomeTaxTree
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a taxonomic subtree by taxonomic identifier
  # Using a NCBI Taxonomy ID or name (common or scientific) at any rank, get a subtree filtered for species with assembled genomes
  # @param character  taxon  NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank 
  # @return [V1alpha1Organism]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

