# Automatically generated by openapi-generator (https://openapi-generator.tech)
# Please update as you see appropriate

context("Test GeneApi")

api.instance <- GeneApi$new()

test_that("DownloadGenePackage", {
  # tests for DownloadGenePackage
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a gene dataset by gene ID
  # Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by gene ID.
  # @param array[integer]  gene.ids   
  # @param array[character]  include.annotation.type  Select additional types of annotation to include in the data package.  If unset, no annotation is provided.  (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @param character  filename  Output file name.  (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("DownloadGenePackagePost", {
  # tests for DownloadGenePackagePost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a gene dataset by POST
  # Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by POST.
  # @param V1alpha1GeneDatasetRequest  body   
  # @param character  filename  Output file name.  (optional)
  # @return [data.frame]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDownloadSummaryByAccession", {
  # tests for GeneDownloadSummaryByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene download summary by RefSeq Accession
  # Get gene download summary by RefSeq Accession in a JSON output format.
  # @param array[character]  accessions  RNA or Protein accessions. 
  # @param character  returned.content    (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @return [V1alpha1DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDownloadSummaryById", {
  # tests for GeneDownloadSummaryById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene download summary by GeneID
  # Get a download summary by GeneID in a JSON output format.
  # @param array[integer]  gene.ids   
  # @param character  returned.content    (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @return [V1alpha1DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDownloadSummaryByPost", {
  # tests for GeneDownloadSummaryByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene download summary
  # Get gene download summary in a JSON output format.
  # @param V1alpha1GeneDatasetRequest  body   
  # @return [V1alpha1DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneDownloadSummaryByTaxAndSymbol", {
  # tests for GeneDownloadSummaryByTaxAndSymbol
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene download summary by gene symbol.
  # Get gene download summary by gene symbol in a JSON output format.
  # @param array[character]  symbols   
  # @param character  taxon   
  # @param character  returned.content    (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @return [V1alpha1DownloadSummary]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataByAccession", {
  # tests for GeneMetadataByAccession
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene metadata by RefSeq Accession
  # Get detailed gene metadata by RefSeq Accession in a JSON output format.
  # @param array[character]  accessions  RNA or Protein accessions. 
  # @param character  returned.content    (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @return [V1alpha1GeneMetadata]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataById", {
  # tests for GeneMetadataById
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene metadata by GeneID
  # Get detailed gene metadata by GeneID in a JSON output format.
  # @param array[integer]  gene.ids   
  # @param character  returned.content    (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @return [V1alpha1GeneMetadata]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataByPost", {
  # tests for GeneMetadataByPost
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene metadata
  # Get detailed gene metadata in a JSON output format.
  # @param V1alpha1GeneDatasetRequest  body   
  # @return [V1alpha1GeneMetadata]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneMetadataByTaxAndSymbol", {
  # tests for GeneMetadataByTaxAndSymbol
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get gene metadata by gene symbol.
  # Get detailed gene metadata by gene symbol in a JSON output format.
  # @param array[character]  symbols   
  # @param character  taxon   
  # @param character  returned.content    (optional)
  # @param character  sort.schema.field    (optional)
  # @param character  sort.schema.direction    (optional)
  # @return [V1alpha1GeneMetadata]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneTaxNameQuery", {
  # tests for GeneTaxNameQuery
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Get a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name.
  # This endpoint retrieves a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name of any rank.
  # @param character  taxon.query  NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank 
  # @return [V1alpha1SciNameAndIds]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

test_that("GeneTaxTree", {
  # tests for GeneTaxTree
  # base path: https://api.ncbi.nlm.nih.gov/datasets/v1alpha
  # Retrieve tax tree
  # @param character  taxon  NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank 
  # @return [V1alpha1Organism]

  # uncomment below to test the operation
  #expect_equal(result, "EXPECTED_RESULT")
})

