# NCBI Datasets API
#
# NCBI service to query and download biological sequence data across all domains of life from NCBI databases.
#
# The version of the OpenAPI document: v1alpha
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Gene operations
#' @description ncbi.datasets.Gene
#' @format An \code{R6Class} generator object
#' @field apiClient Handles the client-server communication.
#'
#' @section Methods:
#' \describe{
#' \strong{ DownloadGenePackage } \emph{ Get a gene dataset by gene ID }
#' Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by gene ID.
#'
#' \itemize{
#' \item \emph{ @param } gene.ids list( integer )
#' \item \emph{ @param } include.annotation.type Enum < [FASTA_GENE, FASTA_RNA, FASTA_PROTEIN] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @param } filename character
#'
#'
#' \item status code : 200 | Download selected gene data as a zip file.
#'
#' \item return type : data.frame 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ DownloadGenePackagePost } \emph{ Get a gene dataset by POST }
#' Get a gene dataset including gene, transcript and protein fasta sequence, annotation and metadata by POST.
#'
#' \itemize{
#' \item \emph{ @param } body \link{V1alpha1GeneDatasetRequest}
#' \item \emph{ @param } filename character
#'
#'
#' \item status code : 200 | Download selected gene data as a zip file.
#'
#' \item return type : data.frame 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneDownloadSummaryByAccession } \emph{ Get gene download summary by RefSeq Accession }
#' Get gene download summary by RefSeq Accession in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } accessions list( character )
#' \item \emph{ @param } returned.content Enum < [COMPLETE, IDS_ONLY] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @returnType } \link{V1alpha1DownloadSummary} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1DownloadSummary 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneDownloadSummaryById } \emph{ Get gene download summary by GeneID }
#' Get a download summary by GeneID in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } gene.ids list( integer )
#' \item \emph{ @param } returned.content Enum < [COMPLETE, IDS_ONLY] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @returnType } \link{V1alpha1DownloadSummary} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1DownloadSummary 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneDownloadSummaryByPost } \emph{ Get gene download summary }
#' Get gene download summary in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } body \link{V1alpha1GeneDatasetRequest}
#' \item \emph{ @returnType } \link{V1alpha1DownloadSummary} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1DownloadSummary 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneDownloadSummaryByTaxAndSymbol } \emph{ Get gene download summary by gene symbol. }
#' Get gene download summary by gene symbol in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } symbols list( character )
#' \item \emph{ @param } taxon character
#' \item \emph{ @param } returned.content Enum < [COMPLETE, IDS_ONLY] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @returnType } \link{V1alpha1DownloadSummary} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1DownloadSummary 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneMetadataByAccession } \emph{ Get gene metadata by RefSeq Accession }
#' Get detailed gene metadata by RefSeq Accession in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } accessions list( character )
#' \item \emph{ @param } returned.content Enum < [COMPLETE, IDS_ONLY] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @returnType } \link{V1alpha1GeneMetadata} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1GeneMetadata 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneMetadataById } \emph{ Get gene metadata by GeneID }
#' Get detailed gene metadata by GeneID in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } gene.ids list( integer )
#' \item \emph{ @param } returned.content Enum < [COMPLETE, IDS_ONLY] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @returnType } \link{V1alpha1GeneMetadata} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1GeneMetadata 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneMetadataByPost } \emph{ Get gene metadata }
#' Get detailed gene metadata in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } body \link{V1alpha1GeneDatasetRequest}
#' \item \emph{ @returnType } \link{V1alpha1GeneMetadata} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1GeneMetadata 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneMetadataByTaxAndSymbol } \emph{ Get gene metadata by gene symbol. }
#' Get detailed gene metadata by gene symbol in a JSON output format.
#'
#' \itemize{
#' \item \emph{ @param } symbols list( character )
#' \item \emph{ @param } taxon character
#' \item \emph{ @param } returned.content Enum < [COMPLETE, IDS_ONLY] > 
#' \item \emph{ @param } sort.schema.field Enum < [SORT_FIELD_UNSPECIFIED, SORT_FIELD_GENE_ID, SORT_FIELD_GENE_TYPE, SORT_FIELD_GENE_SYMBOL] > 
#' \item \emph{ @param } sort.schema.direction Enum < [SORT_DIRECTION_UNSPECIFIED, SORT_DIRECTION_ASCENDING, SORT_DIRECTION_DESCENDING] > 
#' \item \emph{ @returnType } \link{V1alpha1GeneMetadata} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1GeneMetadata 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneTaxNameQuery } \emph{ Get a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name. }
#' This endpoint retrieves a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name of any rank.
#'
#' \itemize{
#' \item \emph{ @param } taxon.query character
#' \item \emph{ @returnType } \link{V1alpha1SciNameAndIds} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1SciNameAndIds 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ GeneTaxTree } \emph{ Retrieve tax tree }
#' 
#'
#' \itemize{
#' \item \emph{ @param } taxon character
#' \item \emph{ @returnType } \link{V1alpha1Organism} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1Organism 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' }
#'
#'
#' @examples
#' \dontrun{
#' ####################  DownloadGenePackage  ####################
#'
#' library(ncbi.datasets)
#' var.gene.ids <- [56] # array[integer] | 
#' var.include.annotation.type <- ['include.annotation.type_example'] # array[character] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#' var.filename <- 'ncbi_dataset.zip' # character | Output file name.
#'
#' #Get a gene dataset by gene ID
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$DownloadGenePackage(var.gene.ids, include.annotation.type=var.include.annotation.type, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction, filename=var.filename)
#'
#'
#' ####################  DownloadGenePackagePost  ####################
#'
#' library(ncbi.datasets)
#' var.body <- V1alpha1GeneDatasetRequest$new() # V1alpha1GeneDatasetRequest | 
#' var.filename <- 'ncbi_dataset.zip' # character | Output file name.
#'
#' #Get a gene dataset by POST
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$DownloadGenePackagePost(var.body, filename=var.filename)
#'
#'
#' ####################  GeneDownloadSummaryByAccession  ####################
#'
#' library(ncbi.datasets)
#' var.accessions <- ['accessions_example'] # array[character] | RNA or Protein accessions.
#' var.returned.content <- 'COMPLETE' # character | 
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#'
#' #Get gene download summary by RefSeq Accession
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneDownloadSummaryByAccession(var.accessions, returned.content=var.returned.content, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction)
#'
#'
#' ####################  GeneDownloadSummaryById  ####################
#'
#' library(ncbi.datasets)
#' var.gene.ids <- [56] # array[integer] | 
#' var.returned.content <- 'COMPLETE' # character | 
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#'
#' #Get gene download summary by GeneID
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneDownloadSummaryById(var.gene.ids, returned.content=var.returned.content, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction)
#'
#'
#' ####################  GeneDownloadSummaryByPost  ####################
#'
#' library(ncbi.datasets)
#' var.body <- V1alpha1GeneDatasetRequest$new() # V1alpha1GeneDatasetRequest | 
#'
#' #Get gene download summary
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneDownloadSummaryByPost(var.body)
#'
#'
#' ####################  GeneDownloadSummaryByTaxAndSymbol  ####################
#'
#' library(ncbi.datasets)
#' var.symbols <- ['symbols_example'] # array[character] | 
#' var.taxon <- 'taxon_example' # character | 
#' var.returned.content <- 'COMPLETE' # character | 
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#'
#' #Get gene download summary by gene symbol.
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneDownloadSummaryByTaxAndSymbol(var.symbols, var.taxon, returned.content=var.returned.content, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction)
#'
#'
#' ####################  GeneMetadataByAccession  ####################
#'
#' library(ncbi.datasets)
#' var.accessions <- ['accessions_example'] # array[character] | RNA or Protein accessions.
#' var.returned.content <- 'COMPLETE' # character | 
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#'
#' #Get gene metadata by RefSeq Accession
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneMetadataByAccession(var.accessions, returned.content=var.returned.content, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction)
#'
#'
#' ####################  GeneMetadataById  ####################
#'
#' library(ncbi.datasets)
#' var.gene.ids <- [56] # array[integer] | 
#' var.returned.content <- 'COMPLETE' # character | 
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#'
#' #Get gene metadata by GeneID
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneMetadataById(var.gene.ids, returned.content=var.returned.content, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction)
#'
#'
#' ####################  GeneMetadataByPost  ####################
#'
#' library(ncbi.datasets)
#' var.body <- V1alpha1GeneDatasetRequest$new() # V1alpha1GeneDatasetRequest | 
#'
#' #Get gene metadata
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneMetadataByPost(var.body)
#'
#'
#' ####################  GeneMetadataByTaxAndSymbol  ####################
#'
#' library(ncbi.datasets)
#' var.symbols <- ['symbols_example'] # array[character] | 
#' var.taxon <- 'taxon_example' # character | 
#' var.returned.content <- 'COMPLETE' # character | 
#' var.sort.schema.field <- 'SORT_FIELD_UNSPECIFIED' # character | 
#' var.sort.schema.direction <- 'SORT_DIRECTION_UNSPECIFIED' # character | 
#'
#' #Get gene metadata by gene symbol.
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneMetadataByTaxAndSymbol(var.symbols, var.taxon, returned.content=var.returned.content, sort.schema.field=var.sort.schema.field, sort.schema.direction=var.sort.schema.direction)
#'
#'
#' ####################  GeneTaxNameQuery  ####################
#'
#' library(ncbi.datasets)
#' var.taxon.query <- 'taxon.query_example' # character | NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
#'
#' #Get a list of taxonomy names and IDs found in the gene dataset given a partial taxonomic name.
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneTaxNameQuery(var.taxon.query)
#'
#'
#' ####################  GeneTaxTree  ####################
#'
#' library(ncbi.datasets)
#' var.taxon <- 'taxon_example' # character | NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
#'
#' #Retrieve tax tree
#' api.instance <- GeneApi$new()
#'
#' result <- api.instance$GeneTaxTree(var.taxon)
#'
#'
#' }
#' @importFrom R6 R6Class
#' @importFrom caTools base64encode
#' @export
GeneApi <- R6::R6Class(
  'GeneApi',
  public = list(
    apiClient = NULL,
    initialize = function(apiClient){
      if (!missing(apiClient)) {
        self$apiClient <- apiClient
      }
      else {
        self$apiClient <- ApiClient$new()
      }
    },
    DownloadGenePackage = function(gene.ids, include.annotation.type=NULL, sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', filename='ncbi_dataset.zip', ...){
      apiResponse <- self$DownloadGenePackageWithHttpInfo(gene.ids, include.annotation.type, sort.schema.field, sort.schema.direction, filename, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    DownloadGenePackageWithHttpInfo = function(gene.ids, include.annotation.type=NULL, sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', filename='ncbi_dataset.zip', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`gene.ids`)) {
        stop("Missing required parameter `gene.ids`.")
      }

      queryParams['include_annotation_type'] <- include.annotation.type

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      queryParams['filename'] <- filename

      urlPath <- "/gene/id/{gene_ids}/download"
      if (!missing(`gene.ids`)) {
        urlPath <- gsub(paste0("\\{", "gene_ids", "\\}"), URLencode(as.character(`gene.ids`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "data.frame", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    DownloadGenePackagePost = function(body, filename='ncbi_dataset.zip', ...){
      apiResponse <- self$DownloadGenePackagePostWithHttpInfo(body, filename, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    DownloadGenePackagePostWithHttpInfo = function(body, filename='ncbi_dataset.zip', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`body`)) {
        stop("Missing required parameter `body`.")
      }

      queryParams['filename'] <- filename

      if (!missing(`body`)) {
        body <- `body`$toJSONString()
      } else {
        body <- NULL
      }

      urlPath <- "/gene/download"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "POST",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "data.frame", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneDownloadSummaryByAccession = function(accessions, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      apiResponse <- self$GeneDownloadSummaryByAccessionWithHttpInfo(accessions, returned.content, sort.schema.field, sort.schema.direction, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneDownloadSummaryByAccessionWithHttpInfo = function(accessions, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`accessions`)) {
        stop("Missing required parameter `accessions`.")
      }

      queryParams['returned_content'] <- returned.content

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      urlPath <- "/gene/accession/{accessions}/download_summary"
      if (!missing(`accessions`)) {
        urlPath <- gsub(paste0("\\{", "accessions", "\\}"), URLencode(as.character(`accessions`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1DownloadSummary", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneDownloadSummaryById = function(gene.ids, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      apiResponse <- self$GeneDownloadSummaryByIdWithHttpInfo(gene.ids, returned.content, sort.schema.field, sort.schema.direction, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneDownloadSummaryByIdWithHttpInfo = function(gene.ids, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`gene.ids`)) {
        stop("Missing required parameter `gene.ids`.")
      }

      queryParams['returned_content'] <- returned.content

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      urlPath <- "/gene/id/{gene_ids}/download_summary"
      if (!missing(`gene.ids`)) {
        urlPath <- gsub(paste0("\\{", "gene_ids", "\\}"), URLencode(as.character(`gene.ids`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1DownloadSummary", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneDownloadSummaryByPost = function(body, ...){
      apiResponse <- self$GeneDownloadSummaryByPostWithHttpInfo(body, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneDownloadSummaryByPostWithHttpInfo = function(body, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`body`)) {
        stop("Missing required parameter `body`.")
      }

      if (!missing(`body`)) {
        body <- `body`$toJSONString()
      } else {
        body <- NULL
      }

      urlPath <- "/gene/download_summary"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "POST",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1DownloadSummary", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneDownloadSummaryByTaxAndSymbol = function(symbols, taxon, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      apiResponse <- self$GeneDownloadSummaryByTaxAndSymbolWithHttpInfo(symbols, taxon, returned.content, sort.schema.field, sort.schema.direction, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneDownloadSummaryByTaxAndSymbolWithHttpInfo = function(symbols, taxon, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`symbols`)) {
        stop("Missing required parameter `symbols`.")
      }

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      queryParams['returned_content'] <- returned.content

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      urlPath <- "/gene/symbol/{symbols}/taxon/{taxon}/download_summary"
      if (!missing(`symbols`)) {
        urlPath <- gsub(paste0("\\{", "symbols", "\\}"), URLencode(as.character(`symbols`), reserved = TRUE), urlPath)
      }

      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), URLencode(as.character(`taxon`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1DownloadSummary", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneMetadataByAccession = function(accessions, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      apiResponse <- self$GeneMetadataByAccessionWithHttpInfo(accessions, returned.content, sort.schema.field, sort.schema.direction, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneMetadataByAccessionWithHttpInfo = function(accessions, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`accessions`)) {
        stop("Missing required parameter `accessions`.")
      }

      queryParams['returned_content'] <- returned.content

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      urlPath <- "/gene/accession/{accessions}"
      if (!missing(`accessions`)) {
        urlPath <- gsub(paste0("\\{", "accessions", "\\}"), URLencode(as.character(`accessions`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1GeneMetadata", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneMetadataById = function(gene.ids, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      apiResponse <- self$GeneMetadataByIdWithHttpInfo(gene.ids, returned.content, sort.schema.field, sort.schema.direction, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneMetadataByIdWithHttpInfo = function(gene.ids, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`gene.ids`)) {
        stop("Missing required parameter `gene.ids`.")
      }

      queryParams['returned_content'] <- returned.content

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      urlPath <- "/gene/id/{gene_ids}"
      if (!missing(`gene.ids`)) {
        urlPath <- gsub(paste0("\\{", "gene_ids", "\\}"), URLencode(as.character(`gene.ids`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1GeneMetadata", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneMetadataByPost = function(body, ...){
      apiResponse <- self$GeneMetadataByPostWithHttpInfo(body, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneMetadataByPostWithHttpInfo = function(body, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`body`)) {
        stop("Missing required parameter `body`.")
      }

      if (!missing(`body`)) {
        body <- `body`$toJSONString()
      } else {
        body <- NULL
      }

      urlPath <- "/gene"

      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "POST",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1GeneMetadata", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneMetadataByTaxAndSymbol = function(symbols, taxon, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      apiResponse <- self$GeneMetadataByTaxAndSymbolWithHttpInfo(symbols, taxon, returned.content, sort.schema.field, sort.schema.direction, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneMetadataByTaxAndSymbolWithHttpInfo = function(symbols, taxon, returned.content='COMPLETE', sort.schema.field='SORT_FIELD_UNSPECIFIED', sort.schema.direction='SORT_DIRECTION_UNSPECIFIED', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`symbols`)) {
        stop("Missing required parameter `symbols`.")
      }

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      queryParams['returned_content'] <- returned.content

      queryParams['sort_schema.field'] <- sort.schema.field

      queryParams['sort_schema.direction'] <- sort.schema.direction

      urlPath <- "/gene/symbol/{symbols}/taxon/{taxon}"
      if (!missing(`symbols`)) {
        urlPath <- gsub(paste0("\\{", "symbols", "\\}"), URLencode(as.character(`symbols`), reserved = TRUE), urlPath)
      }

      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), URLencode(as.character(`taxon`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1GeneMetadata", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneTaxNameQuery = function(taxon.query, ...){
      apiResponse <- self$GeneTaxNameQueryWithHttpInfo(taxon.query, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneTaxNameQueryWithHttpInfo = function(taxon.query, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`taxon.query`)) {
        stop("Missing required parameter `taxon.query`.")
      }

      urlPath <- "/gene/taxon_suggest/{taxon_query}"
      if (!missing(`taxon.query`)) {
        urlPath <- gsub(paste0("\\{", "taxon_query", "\\}"), URLencode(as.character(`taxon.query`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1SciNameAndIds", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    GeneTaxTree = function(taxon, ...){
      apiResponse <- self$GeneTaxTreeWithHttpInfo(taxon, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    GeneTaxTreeWithHttpInfo = function(taxon, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      urlPath <- "/gene/taxon/{taxon}/tree"
      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), URLencode(as.character(`taxon`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1Organism", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    }
  )
)
