# NCBI Datasets API
#
# NCBI service to query and download biological sequence data across all domains of life from NCBI databases.
#
# The version of the OpenAPI document: v1alpha
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title V1alpha1GeneDescriptor
#' @description V1alpha1GeneDescriptor Class
#' @format An \code{R6Class} generator object
#' @field chromosome See chromosomes character [optional]
#'
#' @field chromosomes  list( character ) [optional]
#'
#' @field common_name  character [optional]
#'
#' @field description  character [optional]
#'
#' @field ensembl_gene_ids  list( character ) [optional]
#'
#' @field gene_id mongo-required character [optional]
#'
#' @field genomic_ranges  list( \link{V1alpha1SeqRangeSet} ) [optional]
#'
#' @field genomic_regions  list( \link{V1alpha1GenomicRegion} ) [optional]
#'
#' @field nomenclature_authority  \link{V1alpha1NomenclatureAuthority} [optional]
#'
#' @field omim_ids  list( character ) [optional]
#'
#' @field orientation  \link{V1alpha1Orientation} [optional]
#'
#' @field proteins  list( \link{V1alpha1Protein} ) [optional]
#'
#' @field reference_standards  list( \link{V1alpha1GenomicRegion} ) [optional]
#'
#' @field replaced_gene_id  character [optional]
#'
#' @field rna_type  \link{GeneDescriptorRnaType} [optional]
#'
#' @field swiss_prot_accessions  list( character ) [optional]
#'
#' @field symbol  character [optional]
#'
#' @field synonyms  list( character ) [optional]
#'
#' @field tax_id  character [optional]
#'
#' @field taxname  character [optional]
#'
#' @field transcripts  list( \link{V1alpha1Transcript} ) [optional]
#'
#' @field type  \link{GeneDescriptorGeneType} [optional]
#'
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
V1alpha1GeneDescriptor <- R6::R6Class(
  'V1alpha1GeneDescriptor',
  public = list(
    `chromosome` = NULL,
    `chromosomes` = NULL,
    `common_name` = NULL,
    `description` = NULL,
    `ensembl_gene_ids` = NULL,
    `gene_id` = NULL,
    `genomic_ranges` = NULL,
    `genomic_regions` = NULL,
    `nomenclature_authority` = NULL,
    `omim_ids` = NULL,
    `orientation` = NULL,
    `proteins` = NULL,
    `reference_standards` = NULL,
    `replaced_gene_id` = NULL,
    `rna_type` = NULL,
    `swiss_prot_accessions` = NULL,
    `symbol` = NULL,
    `synonyms` = NULL,
    `tax_id` = NULL,
    `taxname` = NULL,
    `transcripts` = NULL,
    `type` = NULL,
    initialize = function(`chromosome`=NULL, `chromosomes`=NULL, `common_name`=NULL, `description`=NULL, `ensembl_gene_ids`=NULL, `gene_id`=NULL, `genomic_ranges`=NULL, `genomic_regions`=NULL, `nomenclature_authority`=NULL, `omim_ids`=NULL, `orientation`=NULL, `proteins`=NULL, `reference_standards`=NULL, `replaced_gene_id`=NULL, `rna_type`=NULL, `swiss_prot_accessions`=NULL, `symbol`=NULL, `synonyms`=NULL, `tax_id`=NULL, `taxname`=NULL, `transcripts`=NULL, `type`=NULL, ...){
      local.optional.var <- list(...)
      if (!is.null(`chromosome`)) {
        stopifnot(is.character(`chromosome`), length(`chromosome`) == 1)
        self$`chromosome` <- `chromosome`
      }
      if (!is.null(`chromosomes`)) {
        stopifnot(is.vector(`chromosomes`), length(`chromosomes`) != 0)
        sapply(`chromosomes`, function(x) stopifnot(is.character(x)))
        self$`chromosomes` <- `chromosomes`
      }
      if (!is.null(`common_name`)) {
        stopifnot(is.character(`common_name`), length(`common_name`) == 1)
        self$`common_name` <- `common_name`
      }
      if (!is.null(`description`)) {
        stopifnot(is.character(`description`), length(`description`) == 1)
        self$`description` <- `description`
      }
      if (!is.null(`ensembl_gene_ids`)) {
        stopifnot(is.vector(`ensembl_gene_ids`), length(`ensembl_gene_ids`) != 0)
        sapply(`ensembl_gene_ids`, function(x) stopifnot(is.character(x)))
        self$`ensembl_gene_ids` <- `ensembl_gene_ids`
      }
      if (!is.null(`gene_id`)) {
        stopifnot(is.character(`gene_id`), length(`gene_id`) == 1)
        self$`gene_id` <- `gene_id`
      }
      if (!is.null(`genomic_ranges`)) {
        stopifnot(is.vector(`genomic_ranges`), length(`genomic_ranges`) != 0)
        sapply(`genomic_ranges`, function(x) stopifnot(R6::is.R6(x)))
        self$`genomic_ranges` <- `genomic_ranges`
      }
      if (!is.null(`genomic_regions`)) {
        stopifnot(is.vector(`genomic_regions`), length(`genomic_regions`) != 0)
        sapply(`genomic_regions`, function(x) stopifnot(R6::is.R6(x)))
        self$`genomic_regions` <- `genomic_regions`
      }
      if (!is.null(`nomenclature_authority`)) {
        stopifnot(R6::is.R6(`nomenclature_authority`))
        self$`nomenclature_authority` <- `nomenclature_authority`
      }
      if (!is.null(`omim_ids`)) {
        stopifnot(is.vector(`omim_ids`), length(`omim_ids`) != 0)
        sapply(`omim_ids`, function(x) stopifnot(is.character(x)))
        self$`omim_ids` <- `omim_ids`
      }
      if (!is.null(`orientation`)) {
        stopifnot(R6::is.R6(`orientation`))
        self$`orientation` <- `orientation`
      }
      if (!is.null(`proteins`)) {
        stopifnot(is.vector(`proteins`), length(`proteins`) != 0)
        sapply(`proteins`, function(x) stopifnot(R6::is.R6(x)))
        self$`proteins` <- `proteins`
      }
      if (!is.null(`reference_standards`)) {
        stopifnot(is.vector(`reference_standards`), length(`reference_standards`) != 0)
        sapply(`reference_standards`, function(x) stopifnot(R6::is.R6(x)))
        self$`reference_standards` <- `reference_standards`
      }
      if (!is.null(`replaced_gene_id`)) {
        stopifnot(is.character(`replaced_gene_id`), length(`replaced_gene_id`) == 1)
        self$`replaced_gene_id` <- `replaced_gene_id`
      }
      if (!is.null(`rna_type`)) {
        stopifnot(R6::is.R6(`rna_type`))
        self$`rna_type` <- `rna_type`
      }
      if (!is.null(`swiss_prot_accessions`)) {
        stopifnot(is.vector(`swiss_prot_accessions`), length(`swiss_prot_accessions`) != 0)
        sapply(`swiss_prot_accessions`, function(x) stopifnot(is.character(x)))
        self$`swiss_prot_accessions` <- `swiss_prot_accessions`
      }
      if (!is.null(`symbol`)) {
        stopifnot(is.character(`symbol`), length(`symbol`) == 1)
        self$`symbol` <- `symbol`
      }
      if (!is.null(`synonyms`)) {
        stopifnot(is.vector(`synonyms`), length(`synonyms`) != 0)
        sapply(`synonyms`, function(x) stopifnot(is.character(x)))
        self$`synonyms` <- `synonyms`
      }
      if (!is.null(`tax_id`)) {
        stopifnot(is.character(`tax_id`), length(`tax_id`) == 1)
        self$`tax_id` <- `tax_id`
      }
      if (!is.null(`taxname`)) {
        stopifnot(is.character(`taxname`), length(`taxname`) == 1)
        self$`taxname` <- `taxname`
      }
      if (!is.null(`transcripts`)) {
        stopifnot(is.vector(`transcripts`), length(`transcripts`) != 0)
        sapply(`transcripts`, function(x) stopifnot(R6::is.R6(x)))
        self$`transcripts` <- `transcripts`
      }
      if (!is.null(`type`)) {
        stopifnot(R6::is.R6(`type`))
        self$`type` <- `type`
      }
    },
    toJSON = function() {
      V1alpha1GeneDescriptorObject <- list()
      if (!is.null(self$`chromosome`)) {
        V1alpha1GeneDescriptorObject[['chromosome']] <-
          self$`chromosome`
      }
      if (!is.null(self$`chromosomes`)) {
        V1alpha1GeneDescriptorObject[['chromosomes']] <-
          self$`chromosomes`
      }
      if (!is.null(self$`common_name`)) {
        V1alpha1GeneDescriptorObject[['common_name']] <-
          self$`common_name`
      }
      if (!is.null(self$`description`)) {
        V1alpha1GeneDescriptorObject[['description']] <-
          self$`description`
      }
      if (!is.null(self$`ensembl_gene_ids`)) {
        V1alpha1GeneDescriptorObject[['ensembl_gene_ids']] <-
          self$`ensembl_gene_ids`
      }
      if (!is.null(self$`gene_id`)) {
        V1alpha1GeneDescriptorObject[['gene_id']] <-
          self$`gene_id`
      }
      if (!is.null(self$`genomic_ranges`)) {
        V1alpha1GeneDescriptorObject[['genomic_ranges']] <-
          lapply(self$`genomic_ranges`, function(x) x$toJSON())
      }
      if (!is.null(self$`genomic_regions`)) {
        V1alpha1GeneDescriptorObject[['genomic_regions']] <-
          lapply(self$`genomic_regions`, function(x) x$toJSON())
      }
      if (!is.null(self$`nomenclature_authority`)) {
        V1alpha1GeneDescriptorObject[['nomenclature_authority']] <-
          self$`nomenclature_authority`$toJSON()
      }
      if (!is.null(self$`omim_ids`)) {
        V1alpha1GeneDescriptorObject[['omim_ids']] <-
          self$`omim_ids`
      }
      if (!is.null(self$`orientation`)) {
        V1alpha1GeneDescriptorObject[['orientation']] <-
          self$`orientation`$toJSON()
      }
      if (!is.null(self$`proteins`)) {
        V1alpha1GeneDescriptorObject[['proteins']] <-
          lapply(self$`proteins`, function(x) x$toJSON())
      }
      if (!is.null(self$`reference_standards`)) {
        V1alpha1GeneDescriptorObject[['reference_standards']] <-
          lapply(self$`reference_standards`, function(x) x$toJSON())
      }
      if (!is.null(self$`replaced_gene_id`)) {
        V1alpha1GeneDescriptorObject[['replaced_gene_id']] <-
          self$`replaced_gene_id`
      }
      if (!is.null(self$`rna_type`)) {
        V1alpha1GeneDescriptorObject[['rna_type']] <-
          self$`rna_type`$toJSON()
      }
      if (!is.null(self$`swiss_prot_accessions`)) {
        V1alpha1GeneDescriptorObject[['swiss_prot_accessions']] <-
          self$`swiss_prot_accessions`
      }
      if (!is.null(self$`symbol`)) {
        V1alpha1GeneDescriptorObject[['symbol']] <-
          self$`symbol`
      }
      if (!is.null(self$`synonyms`)) {
        V1alpha1GeneDescriptorObject[['synonyms']] <-
          self$`synonyms`
      }
      if (!is.null(self$`tax_id`)) {
        V1alpha1GeneDescriptorObject[['tax_id']] <-
          self$`tax_id`
      }
      if (!is.null(self$`taxname`)) {
        V1alpha1GeneDescriptorObject[['taxname']] <-
          self$`taxname`
      }
      if (!is.null(self$`transcripts`)) {
        V1alpha1GeneDescriptorObject[['transcripts']] <-
          lapply(self$`transcripts`, function(x) x$toJSON())
      }
      if (!is.null(self$`type`)) {
        V1alpha1GeneDescriptorObject[['type']] <-
          self$`type`$toJSON()
      }

      V1alpha1GeneDescriptorObject
    },
    fromJSON = function(V1alpha1GeneDescriptorJson) {
      V1alpha1GeneDescriptorObject <- jsonlite::fromJSON(V1alpha1GeneDescriptorJson)
      if (!is.null(V1alpha1GeneDescriptorObject$`chromosome`)) {
        self$`chromosome` <- V1alpha1GeneDescriptorObject$`chromosome`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`chromosomes`)) {
        self$`chromosomes` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`chromosomes`, "array[character]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`common_name`)) {
        self$`common_name` <- V1alpha1GeneDescriptorObject$`common_name`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`description`)) {
        self$`description` <- V1alpha1GeneDescriptorObject$`description`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`ensembl_gene_ids`)) {
        self$`ensembl_gene_ids` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`ensembl_gene_ids`, "array[character]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`gene_id`)) {
        self$`gene_id` <- V1alpha1GeneDescriptorObject$`gene_id`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`genomic_ranges`)) {
        self$`genomic_ranges` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`genomic_ranges`, "array[V1alpha1SeqRangeSet]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`genomic_regions`)) {
        self$`genomic_regions` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`genomic_regions`, "array[V1alpha1GenomicRegion]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`nomenclature_authority`)) {
        nomenclature_authorityObject <- V1alpha1NomenclatureAuthority$new()
        nomenclature_authorityObject$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$nomenclature_authority, auto_unbox = TRUE, digits = NA))
        self$`nomenclature_authority` <- nomenclature_authorityObject
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`omim_ids`)) {
        self$`omim_ids` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`omim_ids`, "array[character]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`orientation`)) {
        orientationObject <- V1alpha1Orientation$new()
        orientationObject$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$orientation, auto_unbox = TRUE, digits = NA))
        self$`orientation` <- orientationObject
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`proteins`)) {
        self$`proteins` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`proteins`, "array[V1alpha1Protein]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`reference_standards`)) {
        self$`reference_standards` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`reference_standards`, "array[V1alpha1GenomicRegion]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`replaced_gene_id`)) {
        self$`replaced_gene_id` <- V1alpha1GeneDescriptorObject$`replaced_gene_id`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`rna_type`)) {
        rna_typeObject <- GeneDescriptorRnaType$new()
        rna_typeObject$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$rna_type, auto_unbox = TRUE, digits = NA))
        self$`rna_type` <- rna_typeObject
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`swiss_prot_accessions`)) {
        self$`swiss_prot_accessions` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`swiss_prot_accessions`, "array[character]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`symbol`)) {
        self$`symbol` <- V1alpha1GeneDescriptorObject$`symbol`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`synonyms`)) {
        self$`synonyms` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`synonyms`, "array[character]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`tax_id`)) {
        self$`tax_id` <- V1alpha1GeneDescriptorObject$`tax_id`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`taxname`)) {
        self$`taxname` <- V1alpha1GeneDescriptorObject$`taxname`
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`transcripts`)) {
        self$`transcripts` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`transcripts`, "array[V1alpha1Transcript]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1GeneDescriptorObject$`type`)) {
        typeObject <- GeneDescriptorGeneType$new()
        typeObject$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$type, auto_unbox = TRUE, digits = NA))
        self$`type` <- typeObject
      }
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`chromosome`)) {
        sprintf(
        '"chromosome":
          "%s"
                ',
        self$`chromosome`
        )},
        if (!is.null(self$`chromosomes`)) {
        sprintf(
        '"chromosomes":
           [%s]
        ',
        paste(unlist(lapply(self$`chromosomes`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`common_name`)) {
        sprintf(
        '"common_name":
          "%s"
                ',
        self$`common_name`
        )},
        if (!is.null(self$`description`)) {
        sprintf(
        '"description":
          "%s"
                ',
        self$`description`
        )},
        if (!is.null(self$`ensembl_gene_ids`)) {
        sprintf(
        '"ensembl_gene_ids":
           [%s]
        ',
        paste(unlist(lapply(self$`ensembl_gene_ids`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`gene_id`)) {
        sprintf(
        '"gene_id":
          "%s"
                ',
        self$`gene_id`
        )},
        if (!is.null(self$`genomic_ranges`)) {
        sprintf(
        '"genomic_ranges":
        [%s]
',
        paste(sapply(self$`genomic_ranges`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`genomic_regions`)) {
        sprintf(
        '"genomic_regions":
        [%s]
',
        paste(sapply(self$`genomic_regions`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`nomenclature_authority`)) {
        sprintf(
        '"nomenclature_authority":
        %s
        ',
        jsonlite::toJSON(self$`nomenclature_authority`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`omim_ids`)) {
        sprintf(
        '"omim_ids":
           [%s]
        ',
        paste(unlist(lapply(self$`omim_ids`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`orientation`)) {
        sprintf(
        '"orientation":
        %s
        ',
        jsonlite::toJSON(self$`orientation`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`proteins`)) {
        sprintf(
        '"proteins":
        [%s]
',
        paste(sapply(self$`proteins`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`reference_standards`)) {
        sprintf(
        '"reference_standards":
        [%s]
',
        paste(sapply(self$`reference_standards`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`replaced_gene_id`)) {
        sprintf(
        '"replaced_gene_id":
          "%s"
                ',
        self$`replaced_gene_id`
        )},
        if (!is.null(self$`rna_type`)) {
        sprintf(
        '"rna_type":
        %s
        ',
        jsonlite::toJSON(self$`rna_type`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`swiss_prot_accessions`)) {
        sprintf(
        '"swiss_prot_accessions":
           [%s]
        ',
        paste(unlist(lapply(self$`swiss_prot_accessions`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`symbol`)) {
        sprintf(
        '"symbol":
          "%s"
                ',
        self$`symbol`
        )},
        if (!is.null(self$`synonyms`)) {
        sprintf(
        '"synonyms":
           [%s]
        ',
        paste(unlist(lapply(self$`synonyms`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`tax_id`)) {
        sprintf(
        '"tax_id":
          "%s"
                ',
        self$`tax_id`
        )},
        if (!is.null(self$`taxname`)) {
        sprintf(
        '"taxname":
          "%s"
                ',
        self$`taxname`
        )},
        if (!is.null(self$`transcripts`)) {
        sprintf(
        '"transcripts":
        [%s]
',
        paste(sapply(self$`transcripts`, function(x) jsonlite::toJSON(x$toJSON(), auto_unbox=TRUE, digits = NA)), collapse=",")
        )},
        if (!is.null(self$`type`)) {
        sprintf(
        '"type":
        %s
        ',
        jsonlite::toJSON(self$`type`$toJSON(), auto_unbox=TRUE, digits = NA)
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(V1alpha1GeneDescriptorJson) {
      V1alpha1GeneDescriptorObject <- jsonlite::fromJSON(V1alpha1GeneDescriptorJson)
      self$`chromosome` <- V1alpha1GeneDescriptorObject$`chromosome`
      self$`chromosomes` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`chromosomes`, "array[character]", loadNamespace("ncbi.datasets"))
      self$`common_name` <- V1alpha1GeneDescriptorObject$`common_name`
      self$`description` <- V1alpha1GeneDescriptorObject$`description`
      self$`ensembl_gene_ids` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`ensembl_gene_ids`, "array[character]", loadNamespace("ncbi.datasets"))
      self$`gene_id` <- V1alpha1GeneDescriptorObject$`gene_id`
      self$`genomic_ranges` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`genomic_ranges`, "array[V1alpha1SeqRangeSet]", loadNamespace("ncbi.datasets"))
      self$`genomic_regions` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`genomic_regions`, "array[V1alpha1GenomicRegion]", loadNamespace("ncbi.datasets"))
      self$`nomenclature_authority` <- V1alpha1NomenclatureAuthority$new()$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$nomenclature_authority, auto_unbox = TRUE, digits = NA))
      self$`omim_ids` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`omim_ids`, "array[character]", loadNamespace("ncbi.datasets"))
      self$`orientation` <- V1alpha1Orientation$new()$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$orientation, auto_unbox = TRUE, digits = NA))
      self$`proteins` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`proteins`, "array[V1alpha1Protein]", loadNamespace("ncbi.datasets"))
      self$`reference_standards` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`reference_standards`, "array[V1alpha1GenomicRegion]", loadNamespace("ncbi.datasets"))
      self$`replaced_gene_id` <- V1alpha1GeneDescriptorObject$`replaced_gene_id`
      self$`rna_type` <- GeneDescriptorRnaType$new()$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$rna_type, auto_unbox = TRUE, digits = NA))
      self$`swiss_prot_accessions` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`swiss_prot_accessions`, "array[character]", loadNamespace("ncbi.datasets"))
      self$`symbol` <- V1alpha1GeneDescriptorObject$`symbol`
      self$`synonyms` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`synonyms`, "array[character]", loadNamespace("ncbi.datasets"))
      self$`tax_id` <- V1alpha1GeneDescriptorObject$`tax_id`
      self$`taxname` <- V1alpha1GeneDescriptorObject$`taxname`
      self$`transcripts` <- ApiClient$new()$deserializeObj(V1alpha1GeneDescriptorObject$`transcripts`, "array[V1alpha1Transcript]", loadNamespace("ncbi.datasets"))
      self$`type` <- GeneDescriptorGeneType$new()$fromJSON(jsonlite::toJSON(V1alpha1GeneDescriptorObject$type, auto_unbox = TRUE, digits = NA))
      self
    }
  )
)
