# NCBI Datasets API
#
# NCBI service to query and download biological sequence data across all domains of life from NCBI databases.
#
# The version of the OpenAPI document: v1alpha
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title V1alpha1AssemblyDatasetDescriptor
#' @description V1alpha1AssemblyDatasetDescriptor Class
#' @format An \code{R6Class} generator object
#' @field annotation_metadata  \link{V1alpha1AnnotationForAssembly} [optional]
#'
#' @field assembly_accession The GenColl assembly accession
(e.g. GCF_000001405.39) character [optional]
#'
#' @field assembly_category  character [optional]
#'
#' @field assembly_level The level of the assembly, e.g. Chromosome, Scaffold, Contig or Complete Genome character [optional]
#'
#' @field chromosomes  list( character ) [optional]
#'
#' @field contig_n50 Minimum contig length needed to cover 50% of the genome integer [optional]
#'
#' @field display_name  character [optional]
#'
#' @field estimated_size Estimated size of all annotation and sequence files in download character [optional]
#'
#' @field org  \link{V1alpha1Organism} [optional]
#'
#' @field seq_length Combined length of all sequence data character [optional]
#'
#' @field submission_date Date the assembly was provided to NCBI character [optional]
#'
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
V1alpha1AssemblyDatasetDescriptor <- R6::R6Class(
  'V1alpha1AssemblyDatasetDescriptor',
  public = list(
    `annotation_metadata` = NULL,
    `assembly_accession` = NULL,
    `assembly_category` = NULL,
    `assembly_level` = NULL,
    `chromosomes` = NULL,
    `contig_n50` = NULL,
    `display_name` = NULL,
    `estimated_size` = NULL,
    `org` = NULL,
    `seq_length` = NULL,
    `submission_date` = NULL,
    initialize = function(`annotation_metadata`=NULL, `assembly_accession`=NULL, `assembly_category`=NULL, `assembly_level`=NULL, `chromosomes`=NULL, `contig_n50`=NULL, `display_name`=NULL, `estimated_size`=NULL, `org`=NULL, `seq_length`=NULL, `submission_date`=NULL, ...){
      local.optional.var <- list(...)
      if (!is.null(`annotation_metadata`)) {
        stopifnot(R6::is.R6(`annotation_metadata`))
        self$`annotation_metadata` <- `annotation_metadata`
      }
      if (!is.null(`assembly_accession`)) {
        stopifnot(is.character(`assembly_accession`), length(`assembly_accession`) == 1)
        self$`assembly_accession` <- `assembly_accession`
      }
      if (!is.null(`assembly_category`)) {
        stopifnot(is.character(`assembly_category`), length(`assembly_category`) == 1)
        self$`assembly_category` <- `assembly_category`
      }
      if (!is.null(`assembly_level`)) {
        stopifnot(is.character(`assembly_level`), length(`assembly_level`) == 1)
        self$`assembly_level` <- `assembly_level`
      }
      if (!is.null(`chromosomes`)) {
        stopifnot(is.vector(`chromosomes`), length(`chromosomes`) != 0)
        sapply(`chromosomes`, function(x) stopifnot(is.character(x)))
        self$`chromosomes` <- `chromosomes`
      }
      if (!is.null(`contig_n50`)) {
        stopifnot(is.numeric(`contig_n50`), length(`contig_n50`) == 1)
        self$`contig_n50` <- `contig_n50`
      }
      if (!is.null(`display_name`)) {
        stopifnot(is.character(`display_name`), length(`display_name`) == 1)
        self$`display_name` <- `display_name`
      }
      if (!is.null(`estimated_size`)) {
        stopifnot(is.character(`estimated_size`), length(`estimated_size`) == 1)
        self$`estimated_size` <- `estimated_size`
      }
      if (!is.null(`org`)) {
        stopifnot(R6::is.R6(`org`))
        self$`org` <- `org`
      }
      if (!is.null(`seq_length`)) {
        stopifnot(is.character(`seq_length`), length(`seq_length`) == 1)
        self$`seq_length` <- `seq_length`
      }
      if (!is.null(`submission_date`)) {
        stopifnot(is.character(`submission_date`), length(`submission_date`) == 1)
        self$`submission_date` <- `submission_date`
      }
    },
    toJSON = function() {
      V1alpha1AssemblyDatasetDescriptorObject <- list()
      if (!is.null(self$`annotation_metadata`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['annotation_metadata']] <-
          self$`annotation_metadata`$toJSON()
      }
      if (!is.null(self$`assembly_accession`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['assembly_accession']] <-
          self$`assembly_accession`
      }
      if (!is.null(self$`assembly_category`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['assembly_category']] <-
          self$`assembly_category`
      }
      if (!is.null(self$`assembly_level`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['assembly_level']] <-
          self$`assembly_level`
      }
      if (!is.null(self$`chromosomes`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['chromosomes']] <-
          self$`chromosomes`
      }
      if (!is.null(self$`contig_n50`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['contig_n50']] <-
          self$`contig_n50`
      }
      if (!is.null(self$`display_name`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['display_name']] <-
          self$`display_name`
      }
      if (!is.null(self$`estimated_size`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['estimated_size']] <-
          self$`estimated_size`
      }
      if (!is.null(self$`org`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['org']] <-
          self$`org`$toJSON()
      }
      if (!is.null(self$`seq_length`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['seq_length']] <-
          self$`seq_length`
      }
      if (!is.null(self$`submission_date`)) {
        V1alpha1AssemblyDatasetDescriptorObject[['submission_date']] <-
          self$`submission_date`
      }

      V1alpha1AssemblyDatasetDescriptorObject
    },
    fromJSON = function(V1alpha1AssemblyDatasetDescriptorJson) {
      V1alpha1AssemblyDatasetDescriptorObject <- jsonlite::fromJSON(V1alpha1AssemblyDatasetDescriptorJson)
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`annotation_metadata`)) {
        annotation_metadataObject <- V1alpha1AnnotationForAssembly$new()
        annotation_metadataObject$fromJSON(jsonlite::toJSON(V1alpha1AssemblyDatasetDescriptorObject$annotation_metadata, auto_unbox = TRUE, digits = NA))
        self$`annotation_metadata` <- annotation_metadataObject
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`assembly_accession`)) {
        self$`assembly_accession` <- V1alpha1AssemblyDatasetDescriptorObject$`assembly_accession`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`assembly_category`)) {
        self$`assembly_category` <- V1alpha1AssemblyDatasetDescriptorObject$`assembly_category`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`assembly_level`)) {
        self$`assembly_level` <- V1alpha1AssemblyDatasetDescriptorObject$`assembly_level`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`chromosomes`)) {
        self$`chromosomes` <- ApiClient$new()$deserializeObj(V1alpha1AssemblyDatasetDescriptorObject$`chromosomes`, "array[character]", loadNamespace("ncbi.datasets"))
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`contig_n50`)) {
        self$`contig_n50` <- V1alpha1AssemblyDatasetDescriptorObject$`contig_n50`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`display_name`)) {
        self$`display_name` <- V1alpha1AssemblyDatasetDescriptorObject$`display_name`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`estimated_size`)) {
        self$`estimated_size` <- V1alpha1AssemblyDatasetDescriptorObject$`estimated_size`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`org`)) {
        orgObject <- V1alpha1Organism$new()
        orgObject$fromJSON(jsonlite::toJSON(V1alpha1AssemblyDatasetDescriptorObject$org, auto_unbox = TRUE, digits = NA))
        self$`org` <- orgObject
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`seq_length`)) {
        self$`seq_length` <- V1alpha1AssemblyDatasetDescriptorObject$`seq_length`
      }
      if (!is.null(V1alpha1AssemblyDatasetDescriptorObject$`submission_date`)) {
        self$`submission_date` <- V1alpha1AssemblyDatasetDescriptorObject$`submission_date`
      }
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`annotation_metadata`)) {
        sprintf(
        '"annotation_metadata":
        %s
        ',
        jsonlite::toJSON(self$`annotation_metadata`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`assembly_accession`)) {
        sprintf(
        '"assembly_accession":
          "%s"
                ',
        self$`assembly_accession`
        )},
        if (!is.null(self$`assembly_category`)) {
        sprintf(
        '"assembly_category":
          "%s"
                ',
        self$`assembly_category`
        )},
        if (!is.null(self$`assembly_level`)) {
        sprintf(
        '"assembly_level":
          "%s"
                ',
        self$`assembly_level`
        )},
        if (!is.null(self$`chromosomes`)) {
        sprintf(
        '"chromosomes":
           [%s]
        ',
        paste(unlist(lapply(self$`chromosomes`, function(x) paste0('"', x, '"'))), collapse=",")
        )},
        if (!is.null(self$`contig_n50`)) {
        sprintf(
        '"contig_n50":
          %d
                ',
        self$`contig_n50`
        )},
        if (!is.null(self$`display_name`)) {
        sprintf(
        '"display_name":
          "%s"
                ',
        self$`display_name`
        )},
        if (!is.null(self$`estimated_size`)) {
        sprintf(
        '"estimated_size":
          "%s"
                ',
        self$`estimated_size`
        )},
        if (!is.null(self$`org`)) {
        sprintf(
        '"org":
        %s
        ',
        jsonlite::toJSON(self$`org`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`seq_length`)) {
        sprintf(
        '"seq_length":
          "%s"
                ',
        self$`seq_length`
        )},
        if (!is.null(self$`submission_date`)) {
        sprintf(
        '"submission_date":
          "%s"
                ',
        self$`submission_date`
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(V1alpha1AssemblyDatasetDescriptorJson) {
      V1alpha1AssemblyDatasetDescriptorObject <- jsonlite::fromJSON(V1alpha1AssemblyDatasetDescriptorJson)
      self$`annotation_metadata` <- V1alpha1AnnotationForAssembly$new()$fromJSON(jsonlite::toJSON(V1alpha1AssemblyDatasetDescriptorObject$annotation_metadata, auto_unbox = TRUE, digits = NA))
      self$`assembly_accession` <- V1alpha1AssemblyDatasetDescriptorObject$`assembly_accession`
      self$`assembly_category` <- V1alpha1AssemblyDatasetDescriptorObject$`assembly_category`
      self$`assembly_level` <- V1alpha1AssemblyDatasetDescriptorObject$`assembly_level`
      self$`chromosomes` <- ApiClient$new()$deserializeObj(V1alpha1AssemblyDatasetDescriptorObject$`chromosomes`, "array[character]", loadNamespace("ncbi.datasets"))
      self$`contig_n50` <- V1alpha1AssemblyDatasetDescriptorObject$`contig_n50`
      self$`display_name` <- V1alpha1AssemblyDatasetDescriptorObject$`display_name`
      self$`estimated_size` <- V1alpha1AssemblyDatasetDescriptorObject$`estimated_size`
      self$`org` <- V1alpha1Organism$new()$fromJSON(jsonlite::toJSON(V1alpha1AssemblyDatasetDescriptorObject$org, auto_unbox = TRUE, digits = NA))
      self$`seq_length` <- V1alpha1AssemblyDatasetDescriptorObject$`seq_length`
      self$`submission_date` <- V1alpha1AssemblyDatasetDescriptorObject$`submission_date`
      self
    }
  )
)
