# NCBI Datasets API
#
# NCBI service to query and download biological sequence data across all domains of life from NCBI databases.
#
# The version of the OpenAPI document: v1alpha
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title V1alpha1Transcript
#' @description V1alpha1Transcript Class
#' @format An \code{R6Class} generator object
#' @field accession_version Transcript accession-version character [optional]
#'
#' @field cds  \link{V1alpha1SeqRangeSet} [optional]
#'
#' @field ensembl_transcript  character [optional]
#'
#' @field exons  \link{V1alpha1SeqRangeSet} [optional]
#'
#' @field genomic_range  \link{V1alpha1SeqRangeSet} [optional]
#'
#' @field length  integer [optional]
#'
#' @field name  character [optional]
#'
#' @field protein  \link{V1alpha1Protein} [optional]
#'
#' @field type  \link{TranscriptTranscriptType} [optional]
#'
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
V1alpha1Transcript <- R6::R6Class(
  'V1alpha1Transcript',
  public = list(
    `accession_version` = NULL,
    `cds` = NULL,
    `ensembl_transcript` = NULL,
    `exons` = NULL,
    `genomic_range` = NULL,
    `length` = NULL,
    `name` = NULL,
    `protein` = NULL,
    `type` = NULL,
    initialize = function(`accession_version`=NULL, `cds`=NULL, `ensembl_transcript`=NULL, `exons`=NULL, `genomic_range`=NULL, `length`=NULL, `name`=NULL, `protein`=NULL, `type`=NULL, ...){
      local.optional.var <- list(...)
      if (!is.null(`accession_version`)) {
        stopifnot(is.character(`accession_version`), length(`accession_version`) == 1)
        self$`accession_version` <- `accession_version`
      }
      if (!is.null(`cds`)) {
        stopifnot(R6::is.R6(`cds`))
        self$`cds` <- `cds`
      }
      if (!is.null(`ensembl_transcript`)) {
        stopifnot(is.character(`ensembl_transcript`), length(`ensembl_transcript`) == 1)
        self$`ensembl_transcript` <- `ensembl_transcript`
      }
      if (!is.null(`exons`)) {
        stopifnot(R6::is.R6(`exons`))
        self$`exons` <- `exons`
      }
      if (!is.null(`genomic_range`)) {
        stopifnot(R6::is.R6(`genomic_range`))
        self$`genomic_range` <- `genomic_range`
      }
      if (!is.null(`length`)) {
        stopifnot(is.numeric(`length`), length(`length`) == 1)
        self$`length` <- `length`
      }
      if (!is.null(`name`)) {
        stopifnot(is.character(`name`), length(`name`) == 1)
        self$`name` <- `name`
      }
      if (!is.null(`protein`)) {
        stopifnot(R6::is.R6(`protein`))
        self$`protein` <- `protein`
      }
      if (!is.null(`type`)) {
        stopifnot(R6::is.R6(`type`))
        self$`type` <- `type`
      }
    },
    toJSON = function() {
      V1alpha1TranscriptObject <- list()
      if (!is.null(self$`accession_version`)) {
        V1alpha1TranscriptObject[['accession_version']] <-
          self$`accession_version`
      }
      if (!is.null(self$`cds`)) {
        V1alpha1TranscriptObject[['cds']] <-
          self$`cds`$toJSON()
      }
      if (!is.null(self$`ensembl_transcript`)) {
        V1alpha1TranscriptObject[['ensembl_transcript']] <-
          self$`ensembl_transcript`
      }
      if (!is.null(self$`exons`)) {
        V1alpha1TranscriptObject[['exons']] <-
          self$`exons`$toJSON()
      }
      if (!is.null(self$`genomic_range`)) {
        V1alpha1TranscriptObject[['genomic_range']] <-
          self$`genomic_range`$toJSON()
      }
      if (!is.null(self$`length`)) {
        V1alpha1TranscriptObject[['length']] <-
          self$`length`
      }
      if (!is.null(self$`name`)) {
        V1alpha1TranscriptObject[['name']] <-
          self$`name`
      }
      if (!is.null(self$`protein`)) {
        V1alpha1TranscriptObject[['protein']] <-
          self$`protein`$toJSON()
      }
      if (!is.null(self$`type`)) {
        V1alpha1TranscriptObject[['type']] <-
          self$`type`$toJSON()
      }

      V1alpha1TranscriptObject
    },
    fromJSON = function(V1alpha1TranscriptJson) {
      V1alpha1TranscriptObject <- jsonlite::fromJSON(V1alpha1TranscriptJson)
      if (!is.null(V1alpha1TranscriptObject$`accession_version`)) {
        self$`accession_version` <- V1alpha1TranscriptObject$`accession_version`
      }
      if (!is.null(V1alpha1TranscriptObject$`cds`)) {
        cdsObject <- V1alpha1SeqRangeSet$new()
        cdsObject$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$cds, auto_unbox = TRUE, digits = NA))
        self$`cds` <- cdsObject
      }
      if (!is.null(V1alpha1TranscriptObject$`ensembl_transcript`)) {
        self$`ensembl_transcript` <- V1alpha1TranscriptObject$`ensembl_transcript`
      }
      if (!is.null(V1alpha1TranscriptObject$`exons`)) {
        exonsObject <- V1alpha1SeqRangeSet$new()
        exonsObject$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$exons, auto_unbox = TRUE, digits = NA))
        self$`exons` <- exonsObject
      }
      if (!is.null(V1alpha1TranscriptObject$`genomic_range`)) {
        genomic_rangeObject <- V1alpha1SeqRangeSet$new()
        genomic_rangeObject$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$genomic_range, auto_unbox = TRUE, digits = NA))
        self$`genomic_range` <- genomic_rangeObject
      }
      if (!is.null(V1alpha1TranscriptObject$`length`)) {
        self$`length` <- V1alpha1TranscriptObject$`length`
      }
      if (!is.null(V1alpha1TranscriptObject$`name`)) {
        self$`name` <- V1alpha1TranscriptObject$`name`
      }
      if (!is.null(V1alpha1TranscriptObject$`protein`)) {
        proteinObject <- V1alpha1Protein$new()
        proteinObject$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$protein, auto_unbox = TRUE, digits = NA))
        self$`protein` <- proteinObject
      }
      if (!is.null(V1alpha1TranscriptObject$`type`)) {
        typeObject <- TranscriptTranscriptType$new()
        typeObject$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$type, auto_unbox = TRUE, digits = NA))
        self$`type` <- typeObject
      }
    },
    toJSONString = function() {
      jsoncontent <- c(
        if (!is.null(self$`accession_version`)) {
        sprintf(
        '"accession_version":
          "%s"
                ',
        self$`accession_version`
        )},
        if (!is.null(self$`cds`)) {
        sprintf(
        '"cds":
        %s
        ',
        jsonlite::toJSON(self$`cds`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`ensembl_transcript`)) {
        sprintf(
        '"ensembl_transcript":
          "%s"
                ',
        self$`ensembl_transcript`
        )},
        if (!is.null(self$`exons`)) {
        sprintf(
        '"exons":
        %s
        ',
        jsonlite::toJSON(self$`exons`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`genomic_range`)) {
        sprintf(
        '"genomic_range":
        %s
        ',
        jsonlite::toJSON(self$`genomic_range`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`length`)) {
        sprintf(
        '"length":
          %d
                ',
        self$`length`
        )},
        if (!is.null(self$`name`)) {
        sprintf(
        '"name":
          "%s"
                ',
        self$`name`
        )},
        if (!is.null(self$`protein`)) {
        sprintf(
        '"protein":
        %s
        ',
        jsonlite::toJSON(self$`protein`$toJSON(), auto_unbox=TRUE, digits = NA)
        )},
        if (!is.null(self$`type`)) {
        sprintf(
        '"type":
        %s
        ',
        jsonlite::toJSON(self$`type`$toJSON(), auto_unbox=TRUE, digits = NA)
        )}
      )
      jsoncontent <- paste(jsoncontent, collapse = ",")
      paste('{', jsoncontent, '}', sep = "")
    },
    fromJSONString = function(V1alpha1TranscriptJson) {
      V1alpha1TranscriptObject <- jsonlite::fromJSON(V1alpha1TranscriptJson)
      self$`accession_version` <- V1alpha1TranscriptObject$`accession_version`
      self$`cds` <- V1alpha1SeqRangeSet$new()$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$cds, auto_unbox = TRUE, digits = NA))
      self$`ensembl_transcript` <- V1alpha1TranscriptObject$`ensembl_transcript`
      self$`exons` <- V1alpha1SeqRangeSet$new()$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$exons, auto_unbox = TRUE, digits = NA))
      self$`genomic_range` <- V1alpha1SeqRangeSet$new()$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$genomic_range, auto_unbox = TRUE, digits = NA))
      self$`length` <- V1alpha1TranscriptObject$`length`
      self$`name` <- V1alpha1TranscriptObject$`name`
      self$`protein` <- V1alpha1Protein$new()$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$protein, auto_unbox = TRUE, digits = NA))
      self$`type` <- TranscriptTranscriptType$new()$fromJSON(jsonlite::toJSON(V1alpha1TranscriptObject$type, auto_unbox = TRUE, digits = NA))
      self
    }
  )
)
