# NCBI Datasets API
#
# NCBI service to query and download biological sequence data across all domains of life from NCBI databases.
#
# The version of the OpenAPI document: v1alpha
# 
# Generated by: https://openapi-generator.tech

#' @docType class
#' @title Virus operations
#' @description ncbi.datasets.Virus
#' @format An \code{R6Class} generator object
#' @field apiClient Handles the client-server communication.
#'
#' @section Methods:
#' \describe{
#' \strong{ Sars2ProteinDownload } \emph{ Download SARS-CoV-2 protein and CDS datasets by protein name }
#' Download a SARS-CoV-2 protein datasets
#'
#' \itemize{
#' \item \emph{ @param } proteins list( character )
#' \item \emph{ @param } refseq.only character
#' \item \emph{ @param } annotated.only character
#' \item \emph{ @param } released.since character
#' \item \emph{ @param } host character
#' \item \emph{ @param } geo.location character
#' \item \emph{ @param } complete.only character
#' \item \emph{ @param } include.annotation.type Enum < [CDS_FASTA, PROT_FASTA, GENOME_GPFF, PDB_FILES] > 
#' \item \emph{ @param } filename character
#'
#'
#' \item status code : 200 | Download selected SARS-CoV-2 proteins and associated annotation data as a zip file.
#'
#' \item return type : data.frame 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ Sars2ProteinSummary } \emph{ Summary of SARS-CoV-2 protein and CDS datasets by protein name }
#' Download a summary of available SARS-CoV-2 protein datasets
#'
#' \itemize{
#' \item \emph{ @param } proteins list( character )
#' \item \emph{ @param } refseq.only character
#' \item \emph{ @param } annotated.only character
#' \item \emph{ @param } released.since character
#' \item \emph{ @param } host character
#' \item \emph{ @param } geo.location character
#' \item \emph{ @param } complete.only character
#' \item \emph{ @param } include.annotation.type Enum < [CDS_FASTA, PROT_FASTA, GENOME_GPFF, PDB_FILES] > 
#' \item \emph{ @param } format Enum < [tsv, csv, jsonl] > 
#' \item \emph{ @returnType } \link{V1alpha1DownloadSummary} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1DownloadSummary 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ Sars2ProteinTable } \emph{ Get SARS-CoV-2 protein metadata in a tabular format. }
#' Get protein metadata in tabular format for SARS-CoV-2 genomes.
#'
#' \itemize{
#' \item \emph{ @param } proteins list( character )
#' \item \emph{ @param } refseq.only character
#' \item \emph{ @param } annotated.only character
#' \item \emph{ @param } released.since character
#' \item \emph{ @param } host character
#' \item \emph{ @param } geo.location character
#' \item \emph{ @param } complete.only character
#' \item \emph{ @param } table.fields Enum < [nucleotide_accession, species_tax_id, species_name, genus, family, nucleotide_length, isolate_name, sequence_type, nuc_completeness, geo_location, us_state, host_name, host_tax_id, collection_date, bioproject, biosample, polyprotein_name, protein_name, protein_accession, protein_synonym, cds_span] > 
#' \item \emph{ @param } format Enum < [tsv, csv, jsonl] > 
#' \item \emph{ @param } filename character
#' \item \emph{ @returnType } \link{StreamResultOfV1alpha1TabularOutput} \cr
#'
#'
#' \item status code : 200 | A successful response.(streaming responses)
#'
#' \item return type : StreamResultOfV1alpha1TabularOutput 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ VirusGenomeDownload } \emph{ Download Coronavirus genome datasets by taxon }
#' Download a Coronavirus genome datasets by taxon
#'
#' \itemize{
#' \item \emph{ @param } taxon character
#' \item \emph{ @param } refseq.only character
#' \item \emph{ @param } annotated.only character
#' \item \emph{ @param } released.since character
#' \item \emph{ @param } host character
#' \item \emph{ @param } geo.location character
#' \item \emph{ @param } complete.only character
#' \item \emph{ @param } exclude.sequence character
#' \item \emph{ @param } include.annotation.type Enum < [DEFAULT, CDS_FASTA, PROT_FASTA, GENOME_GBFF, GENOME_GPFF, PDB_FILES] > 
#' \item \emph{ @param } filename character
#'
#'
#' \item status code : 200 | Download selected viral genome and associated annotation data as a zip file.
#'
#' \item return type : data.frame 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ VirusGenomeSummary } \emph{ Get summary data for Coronaviridae genomes by taxon }
#' Get summary data and download by command line instructions for Coronaviridae genomes by taxon.
#'
#' \itemize{
#' \item \emph{ @param } taxon character
#' \item \emph{ @param } refseq.only character
#' \item \emph{ @param } annotated.only character
#' \item \emph{ @param } released.since character
#' \item \emph{ @param } host character
#' \item \emph{ @param } geo.location character
#' \item \emph{ @param } complete.only character
#' \item \emph{ @param } exclude.sequence character
#' \item \emph{ @param } include.annotation.type Enum < [DEFAULT, CDS_FASTA, PROT_FASTA, GENOME_GBFF, GENOME_GPFF, PDB_FILES] > 
#' \item \emph{ @returnType } \link{V1alpha1DownloadSummary} \cr
#'
#'
#' \item status code : 200 | A successful response.
#'
#' \item return type : V1alpha1DownloadSummary 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' \strong{ VirusGenomeTable } \emph{ Get viral genomic metadata in a tabular format. }
#' Get viral genomic metadata in tabular format for Coronaviridae genomes by taxon.
#'
#' \itemize{
#' \item \emph{ @param } taxon character
#' \item \emph{ @param } refseq.only character
#' \item \emph{ @param } annotated.only character
#' \item \emph{ @param } released.since character
#' \item \emph{ @param } host character
#' \item \emph{ @param } geo.location character
#' \item \emph{ @param } complete.only character
#' \item \emph{ @param } table.fields Enum < [nucleotide_accession, species_tax_id, species_name, genus, family, nucleotide_length, isolate_name, sequence_type, nuc_completeness, geo_location, us_state, host_name, host_tax_id, collection_date, bioproject, biosample, polyprotein_name, protein_name, protein_accession, protein_synonym, cds_span] > 
#' \item \emph{ @param } format Enum < [tsv, csv, jsonl] > 
#' \item \emph{ @param } filename character
#' \item \emph{ @returnType } \link{StreamResultOfV1alpha1TabularOutput} \cr
#'
#'
#' \item status code : 200 | A successful response.(streaming responses)
#'
#' \item return type : StreamResultOfV1alpha1TabularOutput 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' \item status code : 0 | An unexpected error response.
#'
#' \item return type : GatewayruntimeError 
#' \item response headers :
#'
#' \tabular{ll}{
#' }
#' }
#'
#' }
#'
#'
#' @examples
#' \dontrun{
#' ####################  Sars2ProteinDownload  ####################
#'
#' library(ncbi.datasets)
#' var.proteins <- ['proteins_example'] # array[character] | Which proteins to retrieve in the data package
#' var.refseq.only <- 'refseq.only_example' # character | If true, limit results to RefSeq genomes.
#' var.annotated.only <- 'annotated.only_example' # character | If true, limit results to annotated genomes.
#' var.released.since <- 'released.since_example' # character | If set, limit results to viral genomes that have been released after a specified date and time. April 1, 2020 midnight UTC should be formatted as follows: 2020-04-01T00:00:00.000Z.
#' var.host <- 'host_example' # character | If set, limit results to genomes extracted from this host (Taxonomy ID or name) or its taxonomic descendants.
#' var.geo.location <- 'geo.location_example' # character | Assemblies from this location (country and state, or continent).
#' var.complete.only <- 'complete.only_example' # character | only include complete genomes.
#' var.include.annotation.type <- ['include.annotation.type_example'] # array[character] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
#' var.filename <- 'ncbi_dataset.zip' # character | Output file name.
#'
#' #Download SARS-CoV-2 protein and CDS datasets by protein name
#' api.instance <- VirusApi$new()
#'
#' result <- api.instance$Sars2ProteinDownload(var.proteins, refseq.only=var.refseq.only, annotated.only=var.annotated.only, released.since=var.released.since, host=var.host, geo.location=var.geo.location, complete.only=var.complete.only, include.annotation.type=var.include.annotation.type, filename=var.filename)
#'
#'
#' ####################  Sars2ProteinSummary  ####################
#'
#' library(ncbi.datasets)
#' var.proteins <- ['proteins_example'] # array[character] | Which proteins to retrieve in the data package
#' var.refseq.only <- 'refseq.only_example' # character | If true, limit results to RefSeq genomes.
#' var.annotated.only <- 'annotated.only_example' # character | If true, limit results to annotated genomes.
#' var.released.since <- 'released.since_example' # character | If set, limit results to viral genomes that have been released after a specified date and time. April 1, 2020 midnight UTC should be formatted as follows: 2020-04-01T00:00:00.000Z.
#' var.host <- 'host_example' # character | If set, limit results to genomes extracted from this host (Taxonomy ID or name) or its taxonomic descendants.
#' var.geo.location <- 'geo.location_example' # character | Assemblies from this location (country and state, or continent).
#' var.complete.only <- 'complete.only_example' # character | only include complete genomes.
#' var.include.annotation.type <- ['include.annotation.type_example'] # array[character] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
#' var.format <- 'tsv' # character | Choose download format.
#'
#' #Summary of SARS-CoV-2 protein and CDS datasets by protein name
#' api.instance <- VirusApi$new()
#'
#' result <- api.instance$Sars2ProteinSummary(var.proteins, refseq.only=var.refseq.only, annotated.only=var.annotated.only, released.since=var.released.since, host=var.host, geo.location=var.geo.location, complete.only=var.complete.only, include.annotation.type=var.include.annotation.type, format=var.format)
#'
#'
#' ####################  Sars2ProteinTable  ####################
#'
#' library(ncbi.datasets)
#' var.proteins <- ['proteins_example'] # array[character] | Which proteins to retrieve in the data package
#' var.refseq.only <- 'refseq.only_example' # character | If true, limit results to RefSeq genomes.
#' var.annotated.only <- 'annotated.only_example' # character | If true, limit results to annotated genomes.
#' var.released.since <- 'released.since_example' # character | If set, limit results to viral genomes that have been released after a specified date and time. April 1, 2020 midnight UTC should be formatted as follows: 2020-04-01T00:00:00.000Z.
#' var.host <- 'host_example' # character | If set, limit results to genomes extracted from this host (Taxonomy ID or name) or its taxonomic descendants.
#' var.geo.location <- 'geo.location_example' # character | Assemblies from this location (country and state, or continent).
#' var.complete.only <- 'complete.only_example' # character | only include complete genomes.
#' var.table.fields <- ['table.fields_example'] # array[character] | Specify which fields to include in the tabular report.
#' var.format <- 'tsv' # character | Choose download format.
#' var.filename <- 'ncbi_dataset.zip' # character | Output file name.
#'
#' #Get SARS-CoV-2 protein metadata in a tabular format.
#' api.instance <- VirusApi$new()
#'
#' result <- api.instance$Sars2ProteinTable(var.proteins, refseq.only=var.refseq.only, annotated.only=var.annotated.only, released.since=var.released.since, host=var.host, geo.location=var.geo.location, complete.only=var.complete.only, table.fields=var.table.fields, format=var.format, filename=var.filename)
#'
#'
#' ####################  VirusGenomeDownload  ####################
#'
#' library(ncbi.datasets)
#' var.taxon <- 'taxon_example' # character | NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
#' var.refseq.only <- 'refseq.only_example' # character | If true, limit results to RefSeq genomes.
#' var.annotated.only <- 'annotated.only_example' # character | If true, limit results to annotated genomes.
#' var.released.since <- 'released.since_example' # character | If set, limit results to viral genomes that have been released after a specified date (and optionally, time). April 1, 2020 midnight UTC should be formatted as follows: 2020-04-01T00:00:00.000Z.
#' var.host <- 'host_example' # character | If set, limit results to genomes extracted from this host (Taxonomy ID or name) or its taxonomic descendants.
#' var.geo.location <- 'geo.location_example' # character | Assemblies from this location (country and state, or continent).
#' var.complete.only <- 'complete.only_example' # character | only include complete genomes.
#' var.exclude.sequence <- 'exclude.sequence_example' # character | Set to true to omit the genomic sequence.
#' var.include.annotation.type <- ['include.annotation.type_example'] # array[character] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
#' var.filename <- 'ncbi_dataset.zip' # character | Output file name.
#'
#' #Download Coronavirus genome datasets by taxon
#' api.instance <- VirusApi$new()
#'
#' result <- api.instance$VirusGenomeDownload(var.taxon, refseq.only=var.refseq.only, annotated.only=var.annotated.only, released.since=var.released.since, host=var.host, geo.location=var.geo.location, complete.only=var.complete.only, exclude.sequence=var.exclude.sequence, include.annotation.type=var.include.annotation.type, filename=var.filename)
#'
#'
#' ####################  VirusGenomeSummary  ####################
#'
#' library(ncbi.datasets)
#' var.taxon <- 'taxon_example' # character | NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
#' var.refseq.only <- 'refseq.only_example' # character | If true, limit results to RefSeq genomes.
#' var.annotated.only <- 'annotated.only_example' # character | If true, limit results to annotated genomes.
#' var.released.since <- 'released.since_example' # character | If set, limit results to viral genomes that have been released after a specified date (and optionally, time). April 1, 2020 midnight UTC should be formatted as follows: 2020-04-01T00:00:00.000Z.
#' var.host <- 'host_example' # character | If set, limit results to genomes extracted from this host (Taxonomy ID or name) or its taxonomic descendants.
#' var.geo.location <- 'geo.location_example' # character | Assemblies from this location (country and state, or continent).
#' var.complete.only <- 'complete.only_example' # character | only include complete genomes.
#' var.exclude.sequence <- 'exclude.sequence_example' # character | Set to true to omit the genomic sequence.
#' var.include.annotation.type <- ['include.annotation.type_example'] # array[character] | Select additional types of annotation to include in the data package.  If unset, no annotation is provided.
#'
#' #Get summary data for Coronaviridae genomes by taxon
#' api.instance <- VirusApi$new()
#'
#' result <- api.instance$VirusGenomeSummary(var.taxon, refseq.only=var.refseq.only, annotated.only=var.annotated.only, released.since=var.released.since, host=var.host, geo.location=var.geo.location, complete.only=var.complete.only, exclude.sequence=var.exclude.sequence, include.annotation.type=var.include.annotation.type)
#'
#'
#' ####################  VirusGenomeTable  ####################
#'
#' library(ncbi.datasets)
#' var.taxon <- 'taxon_example' # character | NCBI Taxonomy ID or name (common or scientific) at any taxonomic rank
#' var.refseq.only <- 'refseq.only_example' # character | If true, limit results to RefSeq genomes.
#' var.annotated.only <- 'annotated.only_example' # character | If true, limit results to annotated genomes.
#' var.released.since <- 'released.since_example' # character | If set, limit results to viral genomes that have been released after a specified date (and optionally, time). April 1, 2020 midnight UTC should be formatted as follows: 2020-04-01T00:00:00.000Z.
#' var.host <- 'host_example' # character | If set, limit results to genomes extracted from this host (Taxonomy ID or name) or its taxonomic descendants.
#' var.geo.location <- 'geo.location_example' # character | Assemblies from this location (country and state, or continent).
#' var.complete.only <- 'complete.only_example' # character | only include complete genomes.
#' var.table.fields <- ['table.fields_example'] # array[character] | Specify which fields to include in the tabular report.
#' var.format <- 'tsv' # character | Choose download format.
#' var.filename <- 'ncbi_dataset.zip' # character | Output file name.
#'
#' #Get viral genomic metadata in a tabular format.
#' api.instance <- VirusApi$new()
#'
#' result <- api.instance$VirusGenomeTable(var.taxon, refseq.only=var.refseq.only, annotated.only=var.annotated.only, released.since=var.released.since, host=var.host, geo.location=var.geo.location, complete.only=var.complete.only, table.fields=var.table.fields, format=var.format, filename=var.filename)
#'
#'
#' }
#' @importFrom R6 R6Class
#' @importFrom caTools base64encode
#' @export
VirusApi <- R6::R6Class(
  'VirusApi',
  public = list(
    apiClient = NULL,
    initialize = function(apiClient){
      if (!missing(apiClient)) {
        self$apiClient <- apiClient
      }
      else {
        self$apiClient <- ApiClient$new()
      }
    },
    Sars2ProteinDownload = function(proteins, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, include.annotation.type=NULL, filename='ncbi_dataset.zip', ...){
      apiResponse <- self$Sars2ProteinDownloadWithHttpInfo(proteins, refseq.only, annotated.only, released.since, host, geo.location, complete.only, include.annotation.type, filename, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    Sars2ProteinDownloadWithHttpInfo = function(proteins, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, include.annotation.type=NULL, filename='ncbi_dataset.zip', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`proteins`)) {
        stop("Missing required parameter `proteins`.")
      }

      queryParams['refseq_only'] <- refseq.only

      queryParams['annotated_only'] <- annotated.only

      queryParams['released_since'] <- released.since

      queryParams['host'] <- host

      queryParams['geo_location'] <- geo.location

      queryParams['complete_only'] <- complete.only

      queryParams['include_annotation_type'] <- include.annotation.type

      queryParams['filename'] <- filename

      urlPath <- "/virus/taxon/sars2/protein/{proteins}/download"
      if (!missing(`proteins`)) {
        urlPath <- gsub(paste0("\\{", "proteins", "\\}"), URLencode(as.character(`proteins`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "data.frame", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    Sars2ProteinSummary = function(proteins, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, include.annotation.type=NULL, format='tsv', ...){
      apiResponse <- self$Sars2ProteinSummaryWithHttpInfo(proteins, refseq.only, annotated.only, released.since, host, geo.location, complete.only, include.annotation.type, format, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    Sars2ProteinSummaryWithHttpInfo = function(proteins, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, include.annotation.type=NULL, format='tsv', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`proteins`)) {
        stop("Missing required parameter `proteins`.")
      }

      queryParams['refseq_only'] <- refseq.only

      queryParams['annotated_only'] <- annotated.only

      queryParams['released_since'] <- released.since

      queryParams['host'] <- host

      queryParams['geo_location'] <- geo.location

      queryParams['complete_only'] <- complete.only

      queryParams['include_annotation_type'] <- include.annotation.type

      queryParams['format'] <- format

      urlPath <- "/virus/taxon/sars2/protein/{proteins}"
      if (!missing(`proteins`)) {
        urlPath <- gsub(paste0("\\{", "proteins", "\\}"), URLencode(as.character(`proteins`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1DownloadSummary", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    Sars2ProteinTable = function(proteins, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, table.fields=NULL, format='tsv', filename='ncbi_dataset.zip', ...){
      apiResponse <- self$Sars2ProteinTableWithHttpInfo(proteins, refseq.only, annotated.only, released.since, host, geo.location, complete.only, table.fields, format, filename, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    Sars2ProteinTableWithHttpInfo = function(proteins, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, table.fields=NULL, format='tsv', filename='ncbi_dataset.zip', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`proteins`)) {
        stop("Missing required parameter `proteins`.")
      }

      queryParams['refseq_only'] <- refseq.only

      queryParams['annotated_only'] <- annotated.only

      queryParams['released_since'] <- released.since

      queryParams['host'] <- host

      queryParams['geo_location'] <- geo.location

      queryParams['complete_only'] <- complete.only

      queryParams['table_fields'] <- table.fields

      queryParams['format'] <- format

      queryParams['filename'] <- filename

      urlPath <- "/virus/taxon/sars2/protein/{proteins}/table"
      if (!missing(`proteins`)) {
        urlPath <- gsub(paste0("\\{", "proteins", "\\}"), URLencode(as.character(`proteins`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "StreamResultOfV1alpha1TabularOutput", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    VirusGenomeDownload = function(taxon, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, exclude.sequence=NULL, include.annotation.type=NULL, filename='ncbi_dataset.zip', ...){
      apiResponse <- self$VirusGenomeDownloadWithHttpInfo(taxon, refseq.only, annotated.only, released.since, host, geo.location, complete.only, exclude.sequence, include.annotation.type, filename, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    VirusGenomeDownloadWithHttpInfo = function(taxon, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, exclude.sequence=NULL, include.annotation.type=NULL, filename='ncbi_dataset.zip', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      queryParams['refseq_only'] <- refseq.only

      queryParams['annotated_only'] <- annotated.only

      queryParams['released_since'] <- released.since

      queryParams['host'] <- host

      queryParams['geo_location'] <- geo.location

      queryParams['complete_only'] <- complete.only

      queryParams['exclude_sequence'] <- exclude.sequence

      queryParams['include_annotation_type'] <- include.annotation.type

      queryParams['filename'] <- filename

      urlPath <- "/virus/taxon/{taxon}/genome/download"
      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), URLencode(as.character(`taxon`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "data.frame", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    VirusGenomeSummary = function(taxon, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, exclude.sequence=NULL, include.annotation.type=NULL, ...){
      apiResponse <- self$VirusGenomeSummaryWithHttpInfo(taxon, refseq.only, annotated.only, released.since, host, geo.location, complete.only, exclude.sequence, include.annotation.type, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    VirusGenomeSummaryWithHttpInfo = function(taxon, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, exclude.sequence=NULL, include.annotation.type=NULL, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      queryParams['refseq_only'] <- refseq.only

      queryParams['annotated_only'] <- annotated.only

      queryParams['released_since'] <- released.since

      queryParams['host'] <- host

      queryParams['geo_location'] <- geo.location

      queryParams['complete_only'] <- complete.only

      queryParams['exclude_sequence'] <- exclude.sequence

      queryParams['include_annotation_type'] <- include.annotation.type

      urlPath <- "/virus/taxon/{taxon}/genome"
      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), URLencode(as.character(`taxon`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "V1alpha1DownloadSummary", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    },
    VirusGenomeTable = function(taxon, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, table.fields=NULL, format='tsv', filename='ncbi_dataset.zip', ...){
      apiResponse <- self$VirusGenomeTableWithHttpInfo(taxon, refseq.only, annotated.only, released.since, host, geo.location, complete.only, table.fields, format, filename, ...)
      resp <- apiResponse$response
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        apiResponse$content
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        apiResponse
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        apiResponse
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        apiResponse
      }
    },

    VirusGenomeTableWithHttpInfo = function(taxon, refseq.only=NULL, annotated.only=NULL, released.since=NULL, host=NULL, geo.location=NULL, complete.only=NULL, table.fields=NULL, format='tsv', filename='ncbi_dataset.zip', ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- c()

      if (missing(`taxon`)) {
        stop("Missing required parameter `taxon`.")
      }

      queryParams['refseq_only'] <- refseq.only

      queryParams['annotated_only'] <- annotated.only

      queryParams['released_since'] <- released.since

      queryParams['host'] <- host

      queryParams['geo_location'] <- geo.location

      queryParams['complete_only'] <- complete.only

      queryParams['table_fields'] <- table.fields

      queryParams['format'] <- format

      queryParams['filename'] <- filename

      urlPath <- "/virus/taxon/{taxon}/genome/table"
      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), URLencode(as.character(`taxon`), reserved = TRUE), urlPath)
      }


      resp <- self$apiClient$CallApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)

      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        deserializedRespObj <- tryCatch(
          self$apiClient$deserialize(resp, "StreamResultOfV1alpha1TabularOutput", loadNamespace("ncbi.datasets")),
          error = function(e){
             stop("Failed to deserialize response")
          }
        )
        ApiResponse$new(deserializedRespObj, resp)
      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
        ApiResponse$new(paste("Server returned " , httr::status_code(resp) , " response status code."), resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        ApiResponse$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        ApiResponse$new("API server error", resp)
      }
    }
  )
)
